src/d3a/models/strategy/__init__.py

```diff
@@ -454,14 +455,27 @@ def __init__(self):
         self._bids = {}
         self._traded_bids = {}
 
-    def post_bid(self, market, price, energy, buyer_origin=None):
+    def _remove_existing_bids(self, market: Market) -> None:
+        """Remove all existing bids in the market"""
+
+        for bid in self.get_posted_bids(market):
+            assert bid.buyer == self.owner.name
+            if bid.id in market.bids.keys():
+                bid = market.bids[bid.id]
```

Comment:
Can you please explain what you are trying to achieve here? I wonder how this works, as `bid` is the loop variable.

Reply:
My mistake: this is something I took from another portion of the code and I later forgot to double-check:
https://github.com/gridsingularity/d3a/blob/89e521bf5af64ed05f0dd1e43cc2209df128713e/src/d3a/models/strategy/update_frequency.py#L197-L207

I am actually a bit confused here: what's the difference between `Strategy._bids` and `Market.bids`?

Reply:
You are right, apart from posting bids (or offers) in the markets directly (`Market.bids`), the devices also track the posted and bought/sold bids/offers also in the BaseStrategy (`Strategy._bids`). This enables the strategy to always have knowledge about these information without having to read this from the market. 
Let me know if you need more information. Thanks!

Reply:
In that case, I will leave it as it is 👍 

So the device (`Strategy`) tracks its own bids/offers. Then, at some point, they are posted on the market. Why is there a need to decouple these two lists? 

What I can imagine is something like: the Load has a list of bids (A) that it wants to post. Then the load actually posts them on the market, so the list of market bids (B) is filled with bids from (A). At the same time, (A) is emptied. Could this be the scenario?

Reply:
The list of posted bids in the strategy and in the market should be in sync when posting a bid.
This is the workflow (if you look into `post_bid`)
1. `market.bid` is called which posts the bid to be traded into the market
2. `add_bid_to_posted` is called that keeps track of the bid in the strategy side in order for the strategy to know how much it posted already, because the bid in the market could be instantly traded (and consequently deleted from the market)
Does it make more sense now?

Reply:
Now I understand the process, thank you. My next question is now more abstract: why does the strategy need to know how much it has already posted, if the bid has already been processed by the market?

Reply:
The tracking of bids and offers on the strategy side was used for accounting (producing the results statistics) in earlier times of the d3a. It turned out to be handy when we added the external strategies. Here are two examples:
1. When posting a bid: the strategy can already check the total posted energy on the strategy side without interacting with the market (keep in mind that in the future, the markets will be totally decoupled from the strategies and only accessible via APIs)
2. Let's say the user tries to delete a bid via the external strategy. As the posted bids are tracked in the BaseStrategy class, the ExternalStrategy does not have to read from the market in order to get a list of the bids that have been posted.

But really, I find it hard to find a strong argument for tracking the posted bids and offers also on the strategy side.
@spyrostz Do you know one and help me? Is it maybe time to get rid of this extra accounting?

Reply:
The original rationale behind this was that the markets would eventually end up in the blockchain. Therefore each asset would have to maintain its own copy of the posted bids and offers, in order to avoid requesting the bids / offers from the blockchain, which is potentially a much more expensive and slower operation than keeping a copy of the bids and offers in the strategies. 
Having said that though, we have already decided that even if we decentralize our market classes 100%, we are still going to have one agent for each market in order to avoid such expensive operations. Therefore we will be able to directly access the bids and offers of a market. So this reason is not valid anymore. 
One reason I see in favor of keeping this tracking though, is that in markets with lots of bids and offers, it is going to be much slower to select all the offers and bids from one specific device. For instance, if one market includes 1 million offers, iterating over this list and selecting only the offers for one device would be far more expensive than buffering the offers of each device and having the offers for this device already filtered out. It is premature optimization I know, still I think it is going to be more performant in large markets. 
TLDR I do not have a strong opinion, but I think that it is more optimal to keep the tracking of the bids and offers as it is.

Reply:
Thank you! I think I have a clearer picture now, even though the details of the exact use-case are still a bit abstract 😄 I think it makes sense to keep the two things separate. Adding some more documentation might help to identify more quickly the two separate contexts (market bids vs device bids), so I'll keep this in mind in case I work on those parts of the code.

Reply:
@fievelk sorry, this has not been change. You can delete line 492 and 493. `remove_bid_from_pending` is taking care of the deletion in the market and in the strategy.